### PR TITLE
Process modules in parallel

### PR DESCRIPTION
**Pull request by rayshih**
We implement `processModule` as a `hscFrontendHook` so that we can use
`load` to parallelize haddock build.

(cherry picked from commit 1d89d6a)

### CODE REVIEW

File: haddock-api/src/Haddock/Interface.hs [hunk 1]
```diff
@@ -120,34 +122,55 @@ processModules verbosity modules flags extIfaces = do
...
 --------------------------------------------------------------------------------
 
-
 createIfaces :: Verbosity -> [String] -> [Flag] -> InstIfaceMap -> Ghc ([Interface], ModuleSet)
 createIfaces verbosity modules flags instIfaceMap = do
   -- Ask GHC to tell us what the module graph is
   targets <- mapM (\filePath -> guessTarget filePath Nothing) modules
   setTargets targets
-  modGraph <- depanal [] False
+  _modGraph <- depanal [] False
```
Comment (rayshih) [inline haddock-api/src/Haddock/Interface.hs:130]: I'm not sure why without this line the test won't pass.

I'm also not sure if this breaks the parallelization.

File: haddock-api/src/Haddock/Interface.hs [hunk 2]
```diff
@@ -120,34 +122,55 @@ processModules verbosity modules flags extIfaces = do
...
+                (modifySession $ setHooks $ emptyHooks
+                  { hscFrontendHook = Just $ hook ifaceMapRef
+                  })
+                (modifySession $ setHooks emptyHooks) $
+                GHC.load LoadAllTargets
+  case loadOk of
+    GHC.Succeeded -> do
+      (ifaceMap, !ms) <- liftIO $ atomicModifyIORef' ifaceMapRef $ \i ->
+        ((Map.empty, emptyModuleSet), i)
+      return (Map.elems ifaceMap, ms)
```
Comment (rayshih) [inline haddock-api/src/Haddock/Interface.hs:147]: There was some filtering here, but they seems not required

File: haddock-api/src/Haddock/Interface.hs (hunk 1 above)
Comment (watashi) [reply haddock-api/src/Haddock/Interface.hs:130]: `depanal` will change the state hold by `Ghc` Monad.

File: haddock-api/src/Haddock/Interface.hs (hunk 2 above)
Comment (watashi) [reply haddock-api/src/Haddock/Interface.hs:147]: What "filtering" do you refer to?

Comment (rayshih) [reply haddock-api/src/Haddock/Interface.hs:147]: https://github.com/facebook/fbhaddock/pull/47/commits/e01cccf0773da48ce930d7765a2d733bffed4233#diff-a6119dd2da927bbff2c39eda86cdc733R149-R152

**Comment by josefs** [comment]
I think you want to merge this into the 8.8.3-facebook branch rather than 8.8.2-facebook?

**Comment by rayshih** [comment]
Because there is no ghc-8.8.2 branch in haddock. Is there a better way?

**Comment by josefs** [comment]
Ok. It's probably fine to do it this way.

**Review by josefs — APPROVED** [review]

**Pull request merged**